File: prisma/refs/java/house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        return Math.max(linear(nums, 1, nums.length), linear(nums, 0, nums.length - 1));
    }

    private int linear(int[] nums, int from, int to) {
        int take = 0, skip = 0;
        for (int i = from; i < to; i++) {
            int newTake = skip + nums[i];
            skip = Math.max(take, skip);
            take = newTake;
        }
        return Math.max(take, skip);
    }
}

File: judging/generators/house-robber-ii.ts
import type { Rng } from "./_rng";

export function cases(rng: Rng) {
  const out: { name: string; args: unknown[] }[] = [];
  for (let i = 0; i < 10; i++) {
    out.push({ name: `random circular #${i}`, args: [rng.ints(rng.int(1, 18), 0, 400)] });
  }
  return out;
}

export function stress(rng: Rng) {
  return [{ name: "n=100000", args: [rng.ints(100_000, 0, 400)] }];
}

export function brute(nums: number[]): number {
  if (nums.length === 1) return nums[0];
  const linear = (arr: number[]): number => {
    const go = (i: number): number =>
      i >= arr.length ? 0 : Math.max(go(i + 1), arr[i] + go(i + 2));
    return go(0);
  };
  return Math.max(linear(nums.slice(1)), linear(nums.slice(0, -1)));
}
